# **Airbnb Data Cleaning & Preprocessing**

## Project Objective

The objective of this project is to clean and preprocess a raw Airbnb listings dataset to prepare it for exploratory data analysis. The dataset is inspected for data quality issues such as missing values, duplicate records, inconsistent formatting, incorrect data types, and irrelevant columns. The cleaned dataset will be used for business analysis in the next phase of the project.

## Libraries Used

- Pandas
- NumPy
- Matplotlib

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

Reading the dataset that is to be analysed.

In [2]:
df=pd.read_csv("airbnb_raw_data.csv")

See the top 5 record to actually understand the data.

In [3]:
df.head()

,id,name,rating,reviews,host_name,host_id,address,features,amenities,safety_rules,hourse_rules,img_links,price
0,49849504,Perla bungalov,4.71,64,Entire bungalow hosted by Mehmetcan,357334205.0,"Kartepe, Kocaeli, Turkey","2 guests,2 bedrooms,1 bed,1 bathroom","Mountain view,Valley view,Lake access,Kitchen,...","󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...","Check-in: Flexible,Check out: 12:00 pm,Pets ar...",https://a0.muscache.com/im/pictures/a5da5cb7-c...,"₹8,078 per night"
1,50891766,Authentic Beach Architect Sheltered Villa with...,New,NaN,Entire villa hosted by Fatih,386223873.0,"Kaş, Antalya, Turkey","4 guests,2 bedrooms,2 beds,2 bathrooms","Kitchen,Wifi,Dedicated workspace,Free parking ...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: 4:00 pm - 11:00 pm,Check out: 10:00 ...",https://a0.muscache.com/im/pictures/61b70855-2...,"₹4,665 per night"
2,50699164,cottages sataplia,4.85,68,Entire cabin hosted by Giorgi,409690853.0,"Imereti, Georgia","4 guests,1 bedroom,3 beds,1 bathroom","Mountain view,Kitchen,Wifi,Dedicated workspace...","󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...","Check-in: After 1:00 pm,Check out: 12:00 pm,Se...",https://a0.muscache.com/im/pictures/miso/Hosti...,"₹5,991 per night"
3,49871422,Sapanca Breathable Bungalow,5.0,13,Entire bungalow hosted by Melih,401873242.0,"Sapanca, Sakarya, Turkey","4 guests,1 bedroom,2 beds,1 bathroom","Mountain view,Valley view,Kitchen,Wifi,Free pa...","󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...","Check-in: After 2:00 pm,Check out: 12:00 pm,No...",https://a0.muscache.com/im/pictures/72e6396e-e...,"₹11,339 per night"
4,51245886,Bungalov Ev 2,New,NaN,Entire bungalow hosted by Arp Sapanca,414884116.0,"Sapanca, Sakarya, Turkey","2 guests,1 bedroom,1 bed,1 bathroom","Kitchen,Wifi,Free parking on premises,TV,Air c...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: After 2:00 pm,Check out: 12:00 pm,No...",https://a0.muscache.com/im/pictures/73973308-e...,"₹6,673 per night"


Print random 5 records

In [4]:
df.sample(5,random_state=50)

,id,name,rating,reviews,host_name,host_id,address,features,amenities,safety_rules,hourse_rules,img_links,price
1909,48741462,Hideout In The Apple Orchard Near Kufri|Forest,New,2,Private room in home hosted by Robin,145768978.0,"Fagu, Himachal Pradesh, India","4 guests,2 bedrooms,2 beds,2 bathrooms","Wifi,Free parking on premises,Pets allowed,TV,...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: 12:00 pm - 10:00 pm,Pets are allowed...",https://a0.muscache.com/im/pictures/7dcfa279-c...,"₹7,304 per night"
5183,30217005,Oi-Qaragai Lesnaya Skazka Mountain Resort,New,1,Yurt hosted by Oi-Qaragai Lesnaya Skazka,220033413.0,"Besqaynar, Almaty Province, Kazakhstan","2 guests,1 bedroom,1 bed,1 bathroom","Wifi,Free parking on premises,TV,Hair dryer,Br...","󹀁,Airbnb's COVID-19 safety practices apply,Car...","Check-in: After 2:00 pm,No smoking,No pets,No ...",https://a0.muscache.com/im/pictures/11d19f60-f...,"₹16,143 per night"
2157,592363143976608638,Overnight stay in the site trailer overlooking...,5.0,9,Entire place hosted by Sylvia,451591156.0,"Obergünzburg, Bayern, Germany","2 guests,1 bedroom,1 bed,1 bathroom","Mountain view,Kitchen,Dedicated workspace,Free...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: After 3:00 pm,Pets are allowed",https://a0.muscache.com/im/pictures/miso/Hosti...,"₹2,726 per night"
6854,29914202,"Best location, best views in Gudauri",4.86,7,Entire rental unit hosted by George,10587366.0,"Gudauri, Georgia","4 guests,1 bedroom,1 bed,1 bathroom","Mountain view,Ski-in/Ski-out,Kitchen,Wifi,Free...","󹀁,Airbnb's COVID-19 safety practices apply,Car...","Check-in: After 2:00 pm,Check out: 12:00 pm,Se...",https://a0.muscache.com/im/pictures/3ed3c7e0-0...,"₹7,364 per night"
1068,14781665,Cozy 2 Bhk 🔹Pool🔹Wifi🔹Pwr Backup🔹Kitchen🔹W mac...,4.61,108,Entire serviced apartment hosted by Viren,67706301.0,"Arpora, Goa, India","7 guests,2 bedrooms,2 beds,2 bathrooms","Kitchen,Wifi,Dedicated workspace,Free parking ...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: After 1:00 pm,Check out: 11:00 am,No...",https://a0.muscache.com/im/pictures/1969c0e8-5...,"₹4,930 per night"


### Check dimensionality of the dataset.

In [5]:
df.shape

(14456, 13)

Check non-null count of columns and their datatypes

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14456 entries, 0 to 14455
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            14456 non-null  int64  
 1   name          14455 non-null  object 
 2   rating        14456 non-null  object 
 3   reviews       11712 non-null  object 
 4   host_name     14455 non-null  object 
 5   host_id       14017 non-null  float64
 6   address       14429 non-null  object 
 7   features      14429 non-null  object 
 8   amenities     13467 non-null  object 
 9   safety_rules  14247 non-null  object 
 10  hourse_rules  14247 non-null  object 
 11  img_links     14429 non-null  object 
 12  price         14456 non-null  object 
dtypes: float64(1), int64(1), object(11)
memory usage: 1.4+ MB


### Print stats of all the columns *(including columns of 'object' datatype)*

In [7]:
df.describe(include='all')

,id,name,rating,reviews,host_name,host_id,address,features,amenities,safety_rules,hourse_rules,img_links,price
count,1.445600e+04,14455,14456,11712,14455,1.401700e+04,14429,14429,13467,14247,14247,14429,14456
unique,NaN,13601,77,423,11014,NaN,6575,2694,11630,325,6167,13598,8104
top,NaN,Oops!,New,1,Pension hosted by Onda,NaN,"Manali, Himachal Pradesh, India","2 guests,1 bedroom,1 bed,1 bathroom","Long-term stays allowed,Unavailable: Carbon mo...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...",Check-in: After 3:00 pm,https://a0.muscache.com/im/pictures/miso/Hosti...,"₹2,853 per night"
freq,NaN,26,4841,1318,43,NaN,220,1236,28,3908,217,5,78
mean,7.138227e+16,NaN,NaN,NaN,NaN,1.619384e+08,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,1.962072e+17,NaN,NaN,NaN,NaN,1.403407e+08,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,1.313100e+04,NaN,NaN,NaN,NaN,2.303900e+04,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,2.047078e+07,NaN,NaN,NaN,NaN,3.603925e+07,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,3.524320e+07,NaN,NaN,NaN,NaN,1.243490e+08,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,4.951689e+07,NaN,NaN,NaN,NaN,2.607109e+08,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## Check the number and percentage of missing values for each column.

In [8]:
missing = df.isnull().sum()
missing_percent = (missing/len(df))*100

missing_df=pd.DataFrame({
    "Missing Values" : missing,
    "Percentage" : missing_percent.round(2)
})

print(missing_df)

              Missing Values  Percentage
id                         0        0.00
name                       1        0.01
rating                     0        0.00
reviews                 2744       18.98
host_name                  1        0.01
host_id                  439        3.04
address                   27        0.19
features                  27        0.19
amenities                989        6.84
safety_rules             209        1.45
hourse_rules             209        1.45
img_links                 27        0.19
price                      0        0.00


### Count the number of duplicate values.

In [9]:
df.duplicated().sum()

np.int64(589)

### Check the datatype of each column.

In [10]:
df.dtypes

,0
id,int64
name,object
rating,object
reviews,object
host_name,object
host_id,float64
address,object
features,object
amenities,object
safety_rules,object


### Check the number of unique values in the dataset.

In [11]:
df.nunique().sort_values()

,0
rating,77
safety_rules,325
reviews,423
features,2694
hourse_rules,6167
address,6575
price,8104
host_id,10791
host_name,11014
amenities,11630


### Let's inspect the actual data before proceeding with the data cleaning.

Read the values that are present in the data to get an idea on how to clean it.


In [12]:
df["price"].head()

,price
0,"₹8,078 per night"
1,"₹4,665 per night"
2,"₹5,991 per night"
3,"₹11,339 per night"
4,"₹6,673 per night"


In [13]:
df["rating"].head()

,rating
0,4.71
1,New
2,4.85
3,5.0
4,New


In [14]:
df["address"].head()

,address
0,"Kartepe, Kocaeli, Turkey"
1,"Kaş, Antalya, Turkey"
2,"Imereti, Georgia"
3,"Sapanca, Sakarya, Turkey"
4,"Sapanca, Sakarya, Turkey"


In [15]:
df["amenities"].head()

,amenities
0,"Mountain view,Valley view,Lake access,Kitchen,..."
1,"Kitchen,Wifi,Dedicated workspace,Free parking ..."
2,"Mountain view,Kitchen,Wifi,Dedicated workspace..."
3,"Mountain view,Valley view,Kitchen,Wifi,Free pa..."
4,"Kitchen,Wifi,Free parking on premises,TV,Air c..."


In [16]:
df["features"].head()

,features
0,"2 guests,2 bedrooms,1 bed,1 bathroom"
1,"4 guests,2 bedrooms,2 beds,2 bathrooms"
2,"4 guests,1 bedroom,3 beds,1 bathroom"
3,"4 guests,1 bedroom,2 beds,1 bathroom"
4,"2 guests,1 bedroom,1 bed,1 bathroom"


# Data Cleaning

### Create a copy of the original dataset for cleaning while keeping the original data safe

In [17]:
airbnb = df.copy()

In [18]:
airbnb.columns

Index(['id', 'name', 'rating', 'reviews', 'host_name', 'host_id', 'address',
       'features', 'amenities', 'safety_rules', 'hourse_rules', 'img_links',
       'price'],
      dtype='object')

Rename the column `hourse_rules` to `house_rules` as there is a spelling mistake

In [19]:
airbnb.rename(columns={"hourse_rules":"house_rules"}, inplace=True)

In [20]:
airbnb.columns

Index(['id', 'name', 'rating', 'reviews', 'host_name', 'host_id', 'address',
       'features', 'amenities', 'safety_rules', 'house_rules', 'img_links',
       'price'],
      dtype='object')

#### The column `img_links` has the image links and `id` is only a unique identifier.
So, this 2 columns are not required and hence are removed to simplify the dataset.

In [21]:
airbnb.drop(columns=["id","img_links"], inplace=True)

In [22]:
airbnb.head()

,name,rating,reviews,host_name,host_id,address,features,amenities,safety_rules,house_rules,price
0,Perla bungalov,4.71,64,Entire bungalow hosted by Mehmetcan,357334205.0,"Kartepe, Kocaeli, Turkey","2 guests,2 bedrooms,1 bed,1 bathroom","Mountain view,Valley view,Lake access,Kitchen,...","󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...","Check-in: Flexible,Check out: 12:00 pm,Pets ar...","₹8,078 per night"
1,Authentic Beach Architect Sheltered Villa with...,New,NaN,Entire villa hosted by Fatih,386223873.0,"Kaş, Antalya, Turkey","4 guests,2 bedrooms,2 beds,2 bathrooms","Kitchen,Wifi,Dedicated workspace,Free parking ...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: 4:00 pm - 11:00 pm,Check out: 10:00 ...","₹4,665 per night"
2,cottages sataplia,4.85,68,Entire cabin hosted by Giorgi,409690853.0,"Imereti, Georgia","4 guests,1 bedroom,3 beds,1 bathroom","Mountain view,Kitchen,Wifi,Dedicated workspace...","󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...","Check-in: After 1:00 pm,Check out: 12:00 pm,Se...","₹5,991 per night"
3,Sapanca Breathable Bungalow,5.0,13,Entire bungalow hosted by Melih,401873242.0,"Sapanca, Sakarya, Turkey","4 guests,1 bedroom,2 beds,1 bathroom","Mountain view,Valley view,Kitchen,Wifi,Free pa...","󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...","Check-in: After 2:00 pm,Check out: 12:00 pm,No...","₹11,339 per night"
4,Bungalov Ev 2,New,NaN,Entire bungalow hosted by Arp Sapanca,414884116.0,"Sapanca, Sakarya, Turkey","2 guests,1 bedroom,1 bed,1 bathroom","Kitchen,Wifi,Free parking on premises,TV,Air c...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: After 2:00 pm,Check out: 12:00 pm,No...","₹6,673 per night"


In [23]:
airbnb.isnull().sum()

,0
name,1
rating,0
reviews,2744
host_name,1
host_id,439
address,27
features,27
amenities,989
safety_rules,209
house_rules,209


In [24]:
m1 = airbnb.isnull().sum()
m1[m1>0]

,0
name,1
reviews,2744
host_name,1
host_id,439
address,27
features,27
amenities,989
safety_rules,209
house_rules,209


### Duplicate Records

Identify and removing of Duplicate records to ensure that each Airbnb is represented only once in the dataset.


In [25]:
duplicates = airbnb.duplicated().sum()
print(f"Duplicate Rows: {duplicates}")

Duplicate Rows: 599


Drop the 599 duplicate records found above

In [26]:
airbnb.drop_duplicates(inplace=True)

In [27]:
airbnb.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13857 entries, 0 to 14455
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          13856 non-null  object 
 1   rating        13857 non-null  object 
 2   reviews       11211 non-null  object 
 3   host_name     13856 non-null  object 
 4   host_id       13457 non-null  float64
 5   address       13838 non-null  object 
 6   features      13838 non-null  object 
 7   amenities     12912 non-null  object 
 8   safety_rules  13664 non-null  object 
 9   house_rules   13664 non-null  object 
 10  price         13857 non-null  object 
dtypes: float64(1), object(10)
memory usage: 1.3+ MB


## We need to convert the data type of `price`, `rating` and `reviews` into numeric for performing necessary analysis

*But before converting, we need to need to clean the values properly, so that we don't face value error while converting*

In [28]:
airbnb["price"].unique()[:50]

array(['₹8,078 per night', '₹4,665 per night', '₹5,991 per night',
       '₹11,339 per night', '₹6,673 per night', '₹14,729 per night',
       '₹12,312 per night', '₹13,655 per night', '₹1,747 per night',
       '₹30,486 per night', '₹9,052 per night', '₹5,705 per night',
       '₹14,612 per night', '₹8,929 per night', '₹3,763 per night',
       '₹22,758 per night', '₹11,289 per night', '₹7,526 per night',
       '₹5,018 per night', '₹2,747 per night', '₹6,827 per night',
       '₹5,529 per night', '₹6,924 per night', '₹3,709 per night',
       '₹12,845 per night', '₹8,128 per night', '₹15,052 per night',
       '₹15,604 per night', '₹7,476 per night', '₹1,372 per night',
       '₹14,771 per night', '₹8,248 per night', '₹7,319 per night',
       '₹4,051 per night', '₹14,551 per night', '₹7,418 per night',
       '₹15,975 per night', '₹14,049 per night', '₹17,611 per night',
       '₹13,010 per night', '₹4,059 per night', '₹6,774 per night',
       '₹10,035 per night', '₹8,135 per night

#### Now, we would remove the symbols and unnecessary texts so that, we can convert the datatype to numeric to perform calculation and analysis as needed

In [29]:
airbnb["price"]=(
    airbnb["price"]
    .str.replace("₹","")
    .str.replace(",","")
    .str.replace("per night","")
    .str.replace(" ","")
)

In [30]:
airbnb["price"].unique()[:50]

array(['8078', '4665', '5991', '11339', '6673', '14729', '12312', '13655',
       '1747', '30486', '9052', '5705', '14612', '8929', '3763', '22758',
       '11289', '7526', '5018', '2747', '6827', '5529', '6924', '3709',
       '12845', '8128', '15052', '15604', '7476', '1372', '14771', '8248',
       '7319', '4051', '14551', '7418', '15975', '14049', '17611',
       '13010', '4059', '6774', '10035', '8135', '1712', '8767', '2352',
       '22436', '6372', '19147'], dtype=object)

In [31]:
airbnb['price']=pd.to_numeric(airbnb['price'])

Converting the 'new' or 'New' string values in `rating` column to 'Nan' float values using numpy

In [32]:
airbnb['rating']=airbnb['rating'].replace(['New','NEW'], np.nan)

Now, converting the data type of `rating` to numeric.

In [33]:
airbnb['rating']=pd.to_numeric(airbnb['rating'])

In [34]:
airbnb["reviews"].unique()[:50]

array(['64', nan, '68', '13', '3', '77', '55', '116', '23', '21', '8',
       '52', '17', '61', '49', '38', '4', '15', '26', '5', '2', '85', '9',
       '20', '11', '7', '1', '241', '41', '10', '25', '154', '44', '19',
       '58', '137', '33', '12', '18', '16', '113', '125', '42', '244',
       '27', '188', '34', '69', '97', '6'], dtype=object)

Converting the 'nan' or 'Nan' string values in `reviews` column to 'Nan' float values using numpy

In [35]:
airbnb['reviews']=airbnb['reviews'].replace(['nan','Nan'], np.nan)

In [36]:
airbnb['reviews']= (airbnb["reviews"]
                    .str.replace(",","")
)

In [37]:
airbnb['reviews']=pd.to_numeric(airbnb['reviews'])

In [38]:
airbnb['price'].unique()[:20]

array([ 8078,  4665,  5991, 11339,  6673, 14729, 12312, 13655,  1747,
       30486,  9052,  5705, 14612,  8929,  3763, 22758, 11289,  7526,
        5018,  2747])

In [39]:
airbnb['rating'].unique()[:20]

array([4.71,  nan, 4.85, 5.  , 4.67, 4.97, 4.89, 4.83, 4.87, 4.9 , 4.75,
       4.76, 4.84, 4.79, 4.82, 4.93, 4.96, 4.6 , 4.64, 4.68])

In [40]:
airbnb['reviews'].unique()[:20]

array([ 64.,  nan,  68.,  13.,   3.,  77.,  55., 116.,  23.,  21.,   8.,
        52.,  17.,  61.,  49.,  38.,   4.,  15.,  26.,   5.])

In [41]:
airbnb['reviews'].dtype

dtype('float64')

In [42]:
airbnb.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13857 entries, 0 to 14455
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          13856 non-null  object 
 1   rating        9206 non-null   float64
 2   reviews       11211 non-null  float64
 3   host_name     13856 non-null  object 
 4   host_id       13457 non-null  float64
 5   address       13838 non-null  object 
 6   features      13838 non-null  object 
 7   amenities     12912 non-null  object 
 8   safety_rules  13664 non-null  object 
 9   house_rules   13664 non-null  object 
 10  price         13857 non-null  int64  
dtypes: float64(3), int64(1), object(7)
memory usage: 1.3+ MB


#### Check the columns with data type 'object' now

In [43]:
text_columns = airbnb.select_dtypes(include='object').columns
print(text_columns)

Index(['name', 'host_name', 'address', 'features', 'amenities', 'safety_rules',
       'house_rules'],
      dtype='object')


Remove any extra unnecessary spaces from the data.

In [44]:
for col in text_columns:
  airbnb[col]=airbnb[col].str.strip()

Below is records with duplicate names.

In [45]:
airbnb["name"].duplicated().sum()

np.int64(255)

Check the number and percentage of missing data for each column.

In [46]:
missing1 = airbnb.isnull().sum()
missing_table=pd.DataFrame({
    "Missing values": missing1,
    "Percentage ": ((missing1/len(airbnb))*100).round(2)
})
print(missing_table)

              Missing values  Percentage 
name                       1         0.01
rating                  4651        33.56
reviews                 2646        19.10
host_name                  1         0.01
host_id                  400         2.89
address                   19         0.14
features                  19         0.14
amenities                945         6.82
safety_rules             193         1.39
house_rules              193         1.39
price                      0         0.00


Here, I would like to remove the column `host_id` first, as it is just extra data and would not be helping with any business questions

In [47]:
airbnb.drop(columns="host_id",inplace=True)

In [48]:
airbnb.columns

Index(['name', 'rating', 'reviews', 'host_name', 'address', 'features',
       'amenities', 'safety_rules', 'house_rules', 'price'],
      dtype='object')

### Drop the records with null values in `name`, `host_name`, `address` and `features` columns as their percentage is very small and removing them would not not have any significant changes in our dataset analysis.

In [49]:
airbnb.dropna(
    subset=["name","host_name","address","features"],
    inplace=True
    )

Filling the null values in `amenities`, `safety_rules` and `house_rules` with appropriate string values representing absence of those data.

In [50]:
airbnb["amenities"]=airbnb["amenities"].fillna("No Amenities Listed")
airbnb["safety_rules"]=airbnb["safety_rules"].fillna("No Safety Rules Provided")
airbnb["house_rules"]=airbnb["house_rules"].fillna("No House Rules Provided")

Checking the status of null values now.

In [51]:
airbnb.isnull().sum()

,0
name,0
rating,4640
reviews,2627
host_name,0
address,0
features,0
amenities,0
safety_rules,0
house_rules,0
price,0


In [52]:
airbnb.duplicated().sum()

np.int64(0)

Just in case, if there is any multiple spaces in the dataset. We would be removing that

In [53]:
for col in text_columns:
  airbnb[col] = airbnb[col].str.replace(r"\s+"," ")

In [54]:
airbnb.describe()

,rating,reviews,price
count,9198.000000,11211.000000,1.383800e+04
mean,4.858176,46.385425,2.453747e+04
std,0.139418,73.395248,8.670399e+04
min,3.000000,1.000000,3.930000e+02
25%,4.780000,4.000000,4.089000e+03
50%,4.890000,18.000000,8.516000e+03
75%,4.980000,56.000000,1.734600e+04
max,5.000000,1003.000000,3.906351e+06


In the above stats, we can see that -
 - The min values of `price` and `reviews` is above 0 and hence valid.

 - The min value of `rating` is not less than 0 and max value of `rating` is not more than 5.

Hence, all the values in these columns seems to be possible and hence are valid.

In [55]:
airbnb.head()

,name,rating,reviews,host_name,address,features,amenities,safety_rules,house_rules,price
0,Perla bungalov,4.71,64.0,Entire bungalow hosted by Mehmetcan,"Kartepe, Kocaeli, Turkey","2 guests,2 bedrooms,1 bed,1 bathroom","Mountain view,Valley view,Lake access,Kitchen,...","󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...","Check-in: Flexible,Check out: 12:00 pm,Pets ar...",8078
1,Authentic Beach Architect Sheltered Villa with...,NaN,NaN,Entire villa hosted by Fatih,"Kaş, Antalya, Turkey","4 guests,2 bedrooms,2 beds,2 bathrooms","Kitchen,Wifi,Dedicated workspace,Free parking ...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: 4:00 pm - 11:00 pm,Check out: 10:00 ...",4665
2,cottages sataplia,4.85,68.0,Entire cabin hosted by Giorgi,"Imereti, Georgia","4 guests,1 bedroom,3 beds,1 bathroom","Mountain view,Kitchen,Wifi,Dedicated workspace...","󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...","Check-in: After 1:00 pm,Check out: 12:00 pm,Se...",5991
3,Sapanca Breathable Bungalow,5.00,13.0,Entire bungalow hosted by Melih,"Sapanca, Sakarya, Turkey","4 guests,1 bedroom,2 beds,1 bathroom","Mountain view,Valley view,Kitchen,Wifi,Free pa...","󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...","Check-in: After 2:00 pm,Check out: 12:00 pm,No...",11339
4,Bungalov Ev 2,NaN,NaN,Entire bungalow hosted by Arp Sapanca,"Sapanca, Sakarya, Turkey","2 guests,1 bedroom,1 bed,1 bathroom","Kitchen,Wifi,Free parking on premises,TV,Air c...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: After 2:00 pm,Check out: 12:00 pm,No...",6673


The `features` and `house_rules` columns contain multiple pieces of information in a single text field.

To improve the dataset for analysis, these fields are broken down into separate columns. This will enable more detailed exploratory analysis and business insights.

In [56]:
airbnb['features'].head(10)

,features
0,"2 guests,2 bedrooms,1 bed,1 bathroom"
1,"4 guests,2 bedrooms,2 beds,2 bathrooms"
2,"4 guests,1 bedroom,3 beds,1 bathroom"
3,"4 guests,1 bedroom,2 beds,1 bathroom"
4,"2 guests,1 bedroom,1 bed,1 bathroom"
5,"2 guests,1 bedroom,2 beds,1 bathroom"
6,"2 guests,1 bedroom,1 bed,1 bathroom"
7,"8 guests,1 bedroom,2 beds,1 bathroom"
8,"2 guests,1 bedroom,1 bed,1 bathroom"
9,"10 guests,4 bedrooms,6 beds,3 bathrooms"


In [57]:
airbnb['house_rules'].head(10)

,house_rules
0,"Check-in: Flexible,Check out: 12:00 pm,Pets ar..."
1,"Check-in: 4:00 pm - 11:00 pm,Check out: 10:00 ..."
2,"Check-in: After 1:00 pm,Check out: 12:00 pm,Se..."
3,"Check-in: After 2:00 pm,Check out: 12:00 pm,No..."
4,"Check-in: After 2:00 pm,Check out: 12:00 pm,No..."
5,"Check-in: After 2:00 pm,Check out: 12:00 pm,No..."
6,"Check-in: After 3:00 pm,Check out: 11:00 am,No..."
7,"Check-in: Flexible,Self check-in with building..."
8,"Check-in: After 1:00 pm,Check out: 11:00 am,Se..."
9,"Check-in: After 3:00 pm,No smoking,No pets"


In [58]:
airbnb['guests']=(airbnb['features']
                  .str.extract(r'(\d+)\s*guest?')
                  .astype("Int64")
)

In [59]:
airbnb['bedrooms']=(airbnb['features']
                    .str.extract(r'(\d+)\s*bedroom?')
                    .astype("Int64"))

In [60]:
airbnb['beds']=(airbnb['features']
                .str.extract(r'(\d+)\s*bed?')
                .astype("Int64"))

In [61]:
airbnb['bathrooms']=(airbnb['features']
                     .str.extract(r'(\d+)\s*bathroom?')
                     .astype("Int64"))

We extract the numbers and set the datatype of the columns as 'Int64'

*Note that its not 'int64' but 'Int64' as 'int64' is not capable of containing null values*

In [62]:
airbnb[['features','guests','bedrooms','beds','bathrooms']].head()

,features,guests,bedrooms,beds,bathrooms
0,"2 guests,2 bedrooms,1 bed,1 bathroom",2,2,2,1
1,"4 guests,2 bedrooms,2 beds,2 bathrooms",4,2,2,2
2,"4 guests,1 bedroom,3 beds,1 bathroom",4,1,1,1
3,"4 guests,1 bedroom,2 beds,1 bathroom",4,1,1,1
4,"2 guests,1 bedroom,1 bed,1 bathroom",2,1,1,1


In [63]:
airbnb['check_in']=(airbnb['house_rules']
                    .str.extract(r'Check-in:\s*([^,]+)'))

In [64]:
airbnb['check_out']=(airbnb['house_rules']
                     .str.extract(r'Check out:\s*([^,]+)'))

Similarly, we also extract `check_in` and `check_out` from the `house_rules` column.

Now, we extract the country from the `address` column and store it in a new column.

In [65]:
airbnb['country']=(airbnb['address']
                   .str.split(",")
                   .str[-1]
                   .str.strip()
                   )

In [66]:
airbnb['country'].head(10)

,country
0,Turkey
1,Turkey
2,Georgia
3,Turkey
4,Turkey
5,Turkey
6,Turkey
7,Turkey
8,Vietnam
9,Thailand


In [67]:
airbnb[['house_rules','check_in','check_out']].sample(10)

,house_rules,check_in,check_out
8095,"Check-in: 4:00 pm - 2:00 am,Check out: 11:00 a...",4:00 pm - 2:00 am,11:00 am
10136,"Check-in: 5:00 pm - 8:00 pm,Check out: 10:00 a...",5:00 pm - 8:00 pm,10:00 am
8974,"Check-in: 3:00 pm - 8:00 pm,Check out: 11:00 a...",3:00 pm - 8:00 pm,11:00 am
5232,"Check-in: 7:00 pm - 9:00 pm,No smoking,Pets ar...",7:00 pm - 9:00 pm,NaN
193,"No smoking,Pets are allowed,Show more,Show more,",NaN,NaN
2808,"Check-in: After 3:00 pm,No smoking,Pets are al...",After 3:00 pm,NaN
10124,Check-in: After 2:00 pm,After 2:00 pm,NaN
4837,"Check-in: After 12:00 pm,No pets,󱤇,Smoking is ...",After 12:00 pm,NaN
3446,"Check-in: Flexible,No smoking,No pets,No parti...",Flexible,NaN
12411,"Check-in: After 3:00 pm,Check out: 10:00 am,No...",After 3:00 pm,10:00 am


Now, we remove the feature column as it no longer contains any data that is not extracted and hence is now not useful to us.

In [68]:
airbnb.drop(columns='features',inplace=True)

Now, we remove the unnecessary symbols that is not needed and can probably interfere with the analysis.

In [69]:
airbnb['safety_rules']=airbnb['safety_rules'].str.replace('\U000f1803',"",regex=False)
airbnb['safety_rules']=airbnb['safety_rules'].str.replace(',,',",",regex=False)

In [70]:
row = airbnb.loc[0, "safety_rules"]

for i, ch in enumerate(row):
    print(i, repr(ch))

0 '\U000f9001'
1 ','
2 'A'
3 'i'
4 'r'
5 'b'
6 'n'
7 'b'
8 "'"
9 's'
10 ' '
11 'C'
12 'O'
13 'V'
14 'I'
15 'D'
16 '-'
17 '1'
18 '9'
19 ' '
20 's'
21 'a'
22 'f'
23 'e'
24 't'
25 'y'
26 ' '
27 'p'
28 'r'
29 'a'
30 'c'
31 't'
32 'i'
33 'c'
34 'e'
35 's'
36 ' '
37 'a'
38 'p'
39 'p'
40 'l'
41 'y'
42 ','
43 'N'
44 'o'
45 ' '
46 'c'
47 'a'
48 'r'
49 'b'
50 'o'
51 'n'
52 ' '
53 'm'
54 'o'
55 'n'
56 'o'
57 'x'
58 'i'
59 'd'
60 'e'
61 ' '
62 'a'
63 'l'
64 'a'
65 'r'
66 'm'
67 ','
68 'N'
69 'o'
70 ' '
71 's'
72 'm'
73 'o'
74 'k'
75 'e'
76 ' '
77 'a'
78 'l'
79 'a'
80 'r'
81 'm'
82 ','
83 'N'
84 'e'
85 'a'
86 'r'
87 'b'
88 'y'
89 ' '
90 'l'
91 'a'
92 'k'
93 'e'
94 ','
95 ' '
96 'r'
97 'i'
98 'v'
99 'e'
100 'r'
101 ','
102 ' '
103 'o'
104 't'
105 'h'
106 'e'
107 'r'
108 ' '
109 'b'
110 'o'
111 'd'
112 'y'
113 ' '
114 'o'
115 'f'
116 ' '
117 'w'
118 'a'
119 't'
120 'e'
121 'r'
122 ','
123 'S'
124 'h'
125 'o'
126 'w'
127 ' '
128 'm'
129 'o'
130 'r'
131 'e'
132 ','
133 'S'
134 'h'
135 'o'
136 'w'
137 '

In [71]:
airbnb['safety_rules'].sample(10)

,safety_rules
5094,"󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C..."
4303,"󹀁,Airbnb's COVID-19 safety practices apply,Car..."
5212,"󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C..."
10623,"󹀁,Airbnb's COVID-19 safety practices apply,No ..."
11302,"󹀁,Airbnb's COVID-19 safety practices apply,Car..."
2873,"󹀁,Airbnb's COVID-19 safety practices apply,No ..."
12621,"󹀁,Airbnb's COVID-19 safety practices apply,Sec..."
2354,"󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C..."
7714,"󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C..."
6029,"󹀁,Airbnb's COVID-19 safety practices apply,Car..."


Extract the actual host name from the `host_name` column.

In [72]:
airbnb['host_name']=df['host_name']
airbnb['host_name']=(airbnb['host_name']
                     .str.extract(r'by\s*(.+)'))
airbnb['host_name']

,host_name
0,Mehmetcan
1,Fatih
2,Giorgi
3,Melih
4,Arp Sapanca
...,...
14451,Aidan
14452,Martha
14453,Silas
14454,Cempaka


In [73]:
airbnb

,name,rating,reviews,host_name,address,amenities,safety_rules,house_rules,price,guests,bedrooms,beds,bathrooms,check_in,check_out,country
0,Perla bungalov,4.71,64.0,Mehmetcan,"Kartepe, Kocaeli, Turkey","Mountain view,Valley view,Lake access,Kitchen,...","󹀁,Airbnb's COVID-19 safety practices apply,No ...","Check-in: Flexible,Check out: 12:00 pm,Pets ar...",8078,2,2,2,1,Flexible,12:00 pm,Turkey
1,Authentic Beach Architect Sheltered Villa with...,NaN,NaN,Fatih,"Kaş, Antalya, Turkey","Kitchen,Wifi,Dedicated workspace,Free parking ...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: 4:00 pm - 11:00 pm,Check out: 10:00 ...",4665,4,2,2,2,4:00 pm - 11:00 pm,10:00 am,Turkey
2,cottages sataplia,4.85,68.0,Giorgi,"Imereti, Georgia","Mountain view,Kitchen,Wifi,Dedicated workspace...","󹀁,Airbnb's COVID-19 safety practices apply,No ...","Check-in: After 1:00 pm,Check out: 12:00 pm,Se...",5991,4,1,1,1,After 1:00 pm,12:00 pm,Georgia
3,Sapanca Breathable Bungalow,5.00,13.0,Melih,"Sapanca, Sakarya, Turkey","Mountain view,Valley view,Kitchen,Wifi,Free pa...","󹀁,Airbnb's COVID-19 safety practices apply,No ...","Check-in: After 2:00 pm,Check out: 12:00 pm,No...",11339,4,1,1,1,After 2:00 pm,12:00 pm,Turkey
4,Bungalov Ev 2,NaN,NaN,Arp Sapanca,"Sapanca, Sakarya, Turkey","Kitchen,Wifi,Free parking on premises,TV,Air c...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: After 2:00 pm,Check out: 12:00 pm,No...",6673,2,1,1,1,After 2:00 pm,12:00 pm,Turkey
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14451,"Brown Dome House No. 3, full of healing",4.94,16.0,Aidan,"Imhoe-myeon, Jindo, South Jeolla Province, Sou...","Kitchen,Wifi,Free parking on premises,Private ...","󹀁,Airbnb's COVID-19 safety practices apply,󱠁,C...","Check-in: 3:00 pm - 11:00 pm,Check out: 11:00 ...",4288,4,<NA>,<NA>,8,3:00 pm - 11:00 pm,11:00 am,South Korea
14452,Bubble Hotel Ubud with Bathtub under the Sky,4.67,6.0,Martha,"Gianyar, Indonesia","Wifi,Free on-street parking,Air conditioning,G...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: 3:00 pm - 10:00 pm,Check out: 12:00 ...",15200,2,1,1,<NA>,3:00 pm - 10:00 pm,12:00 pm,Indonesia
14453,Paradise Eco Hub,4.67,3.0,Silas,"Kabale, Uganda","Wifi,Breakfast,Long-term stays allowed,Unavail...","󹀁,Airbnb's COVID-19 safety practices apply,No ...",Check-in: After 3:00 pm,9103,2,1,1,<NA>,After 3:00 pm,NaN,Uganda
14454,WisDOME - Mentigi Bay Village - Gilis - Lombok,4.86,7.0,Cempaka,"Gunung Sari, Nusa Tenggara Barat, Indonesia","Kitchen,Wifi,Dedicated workspace,Free parking ...","󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...","Check-in: Flexible,Check out: 12:00 pm,No pets...",42262,8,2,2,5,Flexible,12:00 pm,Indonesia


In [74]:
airbnb.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13838 entries, 0 to 14455
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          13838 non-null  object 
 1   rating        9198 non-null   float64
 2   reviews       11211 non-null  float64
 3   host_name     13464 non-null  object 
 4   address       13838 non-null  object 
 5   amenities     13838 non-null  object 
 6   safety_rules  13838 non-null  object 
 7   house_rules   13838 non-null  object 
 8   price         13838 non-null  int64  
 9   guests        13838 non-null  Int64  
 10  bedrooms      13118 non-null  Int64  
 11  beds          13795 non-null  Int64  
 12  bathrooms     10388 non-null  Int64  
 13  check_in      13011 non-null  object 
 14  check_out     11312 non-null  object 
 15  country       13838 non-null  object 
dtypes: Int64(4), float64(2), int64(1), object(9)
memory usage: 2.4+ MB


In [75]:
airbnb.to_csv("airbnb_cleaned.csv",index=False)

We saved the new and clean dataset as `airbnb_cleaned.csv`

Now, we would analyze this dataset to get our business insights.

In [76]:
airbnb.shape

(13838, 16)

# Data Cleaning Summary
The raw Airbnb dataset was cleaned and now is ready for analysis.
The following steps were performed:

- Corrected misspelled column names.
- Removed irrelevant columns (`id`, `host_id` and `img_links`) and records with missing important information (`name`, `host_name`, `address`, and `features`).
- Filled missing values in text-based columns with descriptive statements.
- Retained missing values in `rating` and `reviews` as they represent unrated or newly listed properties.
- Removed duplicate records.
- Standardized text formatting by removing any unnecessary whitespaces.
- Extracted the number of guests, bedrooms, beds, and bathrooms from the `features` column.
- Extracted check-in and check-out timings from the `house_rules` column.
- Removed the original `features` column after extracting its information.

### Final Dataset

- Rows: **13,838**
- Columns: **15**
- Output file: **airbnb_cleaned.csv**